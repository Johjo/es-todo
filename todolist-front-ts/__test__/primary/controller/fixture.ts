import {Task, WhichTask} from "@/hexagon/whichTaskQuery/whichTask.query";
import {ChooseAndIgnoreTask} from "@/hexagon/chooseTask/chooseTask.usecase";

export class ChooseTaskForTest implements ChooseAndIgnoreTask.Contract {
    private _history: undefined | [number, number] = undefined;

    execute(chosenTaskId: number, ignoredTaskId: number) : Promise<void> {
        this._history = [chosenTaskId, ignoredTaskId];
        return Promise.resolve();
    }

    history() {
        assert(this._history !== undefined, 'history() called before execute()');
        return this._history;
    }
}

export class WhichTaskQueryForTest implements WhichTask.Contract {
    private _tasks: Task[] = [];

    query(): Promise<Task[]> {
        return Promise.resolve(this._tasks);
    }

    feed(tasks: Task[]) {
        this._tasks = tasks;
    }
}

export class StoreForTest {
    _history: any[] = [];

    history(): any[] {
        return this._history
    }

    dispatch(expectedElement: any) {
        this._history.push(expectedElement);
    }
}

