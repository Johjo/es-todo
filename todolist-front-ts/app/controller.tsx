import {AppStore} from "@/lib/store";
import {todoListFetched} from "@/lib/todolist.slice";
import {createContext, useContext} from "react";
import React from "react";

export type Todolist = { numberOfTasks: number };

export interface TodolistReader {
    onlyOne(): Promise<Todolist>;
}

export interface DependenciesList {
    todolistReaderForRefreshTodolist(): TodolistReader;
}

const DependenciesContext = createContext<DependenciesList | undefined>(undefined)

export const useDependencies = () => {
    const context = useContext(DependenciesContext);
    if (!context) {
        throw new Error("useDependencies doit être utilisé dans un DependenciesProvider");
    }
    return context;
};

export const DependenciesProvider: React.FC<{ dependencies: DependenciesList; children: React.ReactNode }> = ({
                                                                                                                  dependencies,
                                                                                                                  children,
                                                                                                              }) => {
    return <DependenciesContext.Provider value={dependencies}>{children}</DependenciesContext.Provider>;
};

export class Controller {
    constructor(private readonly store: any, private readonly dependencies: DependenciesList) {
    }

    refreshTodolist() {
        const allTodolist = this.dependencies.todolistReaderForRefreshTodolist();

        allTodolist.onlyOne().then(todolist => {
            this.store.dispatch(todoListFetched({numberOfTasks: todolist.numberOfTasks}));
        });

    }
}