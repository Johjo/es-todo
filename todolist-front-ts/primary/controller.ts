import {ChooseAndIgnoreTaskContract, ChooseTaskUseCase, ScreenPort, TodolistPort} from "@/hexagon/chooseTask.usecase";

export class Controller {
    constructor(private readonly dependencies: DependencyList) {

    }

    askForWhichTask(chosenTaskId: number, ignoredTaskId: number) {
        const chooseAndIgnoreTask = this.dependencies.chooseAndIgnoreTaskUseCase();
        chooseAndIgnoreTask.execute(chosenTaskId, ignoredTaskId);
    }
}

export interface DependencyList {
    chooseAndIgnoreTaskUseCase(): ChooseAndIgnoreTaskContract;
}

export abstract class DependencyListOnlyUseCase implements DependencyList {
    chooseAndIgnoreTaskUseCase(): ChooseAndIgnoreTaskContract {
        const todolist = this.todolistForChooseAndIgnoreTaskUseCase();
        const screen = this.screenForChooseAndIgnoreTaskUseCase();
        return new ChooseTaskUseCase(todolist, screen);
    }

    protected abstract todolistForChooseAndIgnoreTaskUseCase(): TodolistPort;

    protected abstract screenForChooseAndIgnoreTaskUseCase(): ScreenPort;
}