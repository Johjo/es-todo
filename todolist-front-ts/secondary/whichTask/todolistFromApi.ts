import {Task, TodolistPort} from "@/hexagon/whichTaskQuery/whichTask.query";


export function build(): TodolistPort {
    return new TodolistFromApi()
}

export class TodolistFromApi implements TodolistPort {
    async whichTask(): Promise<Task[]> {
        const response: Response = await fetch("http://127.0.0.1:8090/rest/todo/Jonathan/which_task");

        const data = await response.json();
        return data.fvpTasks;
    }
}
