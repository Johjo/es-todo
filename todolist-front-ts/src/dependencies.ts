import type { TodolistFetcherPort } from './hexagon/fetchTodolist.port.ts';
import type { OpenTaskUpdaterPort, UuidGeneratorPort } from './hexagon/openTask.port.ts';
import type { DependenciesUseCase } from './dependenciesUseCase.ts';
import type { AppDispatch } from './hexagon/store.ts';
import { FetchTodolistPrimaryPort, FetchTodolistUseCase } from './hexagon/fetchTodolist.usecase.ts';
import { OpenTaskPrimaryPort, OpenTaskUseCase } from './hexagon/openTask.usecase.ts';

export interface DependenciesAdapter {

  todolistFetcher(): TodolistFetcherPort;

  uuidGenerator(): UuidGeneratorPort;

  todolistUpdater(): OpenTaskUpdaterPort;
}

export class DependenciesUseCaseImpl implements DependenciesUseCase {
  constructor(private readonly adapters: DependenciesAdapter) {
  }

  fetchTodolist(dispatch: AppDispatch): FetchTodolistPrimaryPort {
    const todolistFetcher: TodolistFetcherPort = this.adapters.todolistFetcher();
    return new FetchTodolistUseCase(todolistFetcher, dispatch);
  }

  openTask(dispatch: AppDispatch): OpenTaskPrimaryPort {
    const uuidGenerator: UuidGeneratorPort = this.adapters.uuidGenerator();
    const todolistUpdater: OpenTaskUpdaterPort = this.adapters.todolistUpdater();
    return new OpenTaskUseCase(uuidGenerator, todolistUpdater, dispatch);
  }
}
