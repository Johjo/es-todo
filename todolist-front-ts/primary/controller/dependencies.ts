import {ChooseAndIgnoreTask} from "@/hexagon/chooseTask/chooseTask.usecase";
import {WhichTask} from "@/hexagon/whichTaskQuery/whichTask.query";
import {build} from "@/secondary/whichTask/todolistFromApi";

export type Dependencies = WhichTask.Dependencies & ChooseAndIgnoreTask.Dependencies;

function throwNotInjected(message: string) {
    return function (): any {
        throw new Error(message);
    };
}

export const emptyDependencies: Dependencies = {
    chooseAndIgnoreTask: {
        useCase: throwNotInjected("chooseAndIgnoreTask.useCase not injected"),
        adapter: {todolist: throwNotInjected("chooseAndIgnoreTask.adapter.todolist not Injected")},
    }, whichTask: {
        useCase: throwNotInjected(("whichTask.useCase not injected")),
        adapter: {todolist: throwNotInjected("whichTask.adapter.todolist not Injected")},
    }
}

export const allUseCasesDependencies: Dependencies = {
    ...emptyDependencies,
    chooseAndIgnoreTask: {...emptyDependencies.chooseAndIgnoreTask, useCase: ChooseAndIgnoreTask.build},
    whichTask: {...emptyDependencies.whichTask, useCase: WhichTask.build}
}

export const allAdaptersDependencies: Dependencies = {
    ...allUseCasesDependencies,
    whichTask: {...allUseCasesDependencies.whichTask, adapter: {todolist: build}}
}




