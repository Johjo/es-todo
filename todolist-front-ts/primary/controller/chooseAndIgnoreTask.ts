import {WhichTaskUpdated} from "@/lib/todolist.slice";
import {Dependencies} from "@/primary/controller/dependencies";
import {WhichTask} from "@/hexagon/whichTaskQuery/whichTask.query";
import {ChooseAndIgnoreTask} from "@/hexagon/chooseTask/chooseTask.usecase";

export class Controller {
    private _store: StoreContract;
    private _whichTaskQuery: WhichTask.Contract;
    private _chooseAndIgnoreTask: ChooseAndIgnoreTask.Contract;

    constructor(dependencies: Dependencies, store: StoreContract) {
        this._store = store;
        this._whichTaskQuery = dependencies.whichTask.useCase(dependencies);
        this._chooseAndIgnoreTask = dependencies.chooseAndIgnoreTask.useCase(dependencies);
    }

    async chooseAndIgnoreTask(chosenTaskId: number, ignoredTaskId: number) {
        await this._chooseAndIgnoreTask.execute(chosenTaskId, ignoredTaskId);
        const tasks = await this._whichTaskQuery.query();
        this._store.dispatch(WhichTaskUpdated({tasks: tasks}));
    }
}

export interface StoreContract {
    dispatch(whichTaskUpdated: any): void;
}
