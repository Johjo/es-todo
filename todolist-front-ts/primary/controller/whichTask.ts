import {WhichTask} from "@/hexagon/whichTaskQuery/whichTask.query";
import {StoreContract} from "@/primary/controller/chooseAndIgnoreTask";
import {Dependencies} from "@/primary/controller/dependencies";
import {WhichTaskUpdated} from "@/lib/todolist.slice";

export class Controller {
    private _whichTaskQuery: WhichTask.Contract;
    private _store: StoreContract;

    constructor(dependencies: Dependencies, store: StoreContract) {
        this._store = store;
        this._whichTaskQuery = dependencies.whichTask.useCase(dependencies)
    }


    async execute() {
        const tasksToExamine = await this._whichTaskQuery.query();

        this._store.dispatch(WhichTaskUpdated({
            tasks: tasksToExamine
        }));
    }

}