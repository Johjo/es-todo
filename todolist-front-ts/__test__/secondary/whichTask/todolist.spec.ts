import {allAdaptersDependencies} from "@/primary/controller/dependencies";
import {TodolistFromApi} from "@/secondary/whichTask/todolistFromApi";

describe("Todolist", () => {
    it("should return the list of tasks", async () => {
        const sut : TodolistFromApi = allAdaptersDependencies.whichTask.adapter.todolist()
        const tasks = await sut.whichTask();
        expect(tasks).toMatchSnapshot();
    });
});