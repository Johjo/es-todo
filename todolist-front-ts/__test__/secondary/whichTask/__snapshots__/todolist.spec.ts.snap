// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Todolist > should return the list of tasks 1`] = `
[
  {
    "id": 7,
    "name": "#context3",
  },
  {
    "id": 8,
    "name": "Une autre t√¢che #toto",
  },
]
`;
